### ISSUE TITLE
APA date not working properly with bibtex and YAML

### ISSUE THREAD

**Issue by gkourgkoutas** [typst/hayagriva#246]
### Description
When I'm using my bibliography and I choose APA7 as cite style, it doesn't matter if I use a biblatex file or YAML (typst hayagriva). In both cases, if the date (year) is added, there are some formatting issues.

Bibtex file:
```latex
@online{nistCVE,
	author = {NIST},
	title = {CVE-2021-44228},
	url = {https://nvd.nist.gov/vuln/detail/CVE-2021-44228},
        date = {2021},
	urldate = {2024-10-18},
}
```

YAML:
```yaml
nistCVE:
 author: "NIST"
 type: "Web"
 title: "CVE-2021-44228"
 date: "2021"
 url: 
    value: "https://nvd.nist.gov/vuln/detail/CVE-2021-44228"
    date: 2024-10-28
```

These produce the following output:

<img width="885" alt="grafik" src="https://github.com/user-attachments/assets/d02dfb2e-f0a1-4139-b79a-3cc7c3752663">

While the output seems fine, there is no "Retrieved + Date", which is the case in latex when `\urldate` is used.

Note that there is also a comma after the year (2021,) which should not be there if only the year of publication is given. typst expects a full date like 2021-01-01.

If the date (publication year or year in bibtex) is removed...

```latex
@online{nistCVE,
	author = {NIST},
	title = {CVE-2021-44228},
	url = {https://nvd.nist.gov/vuln/detail/CVE-2021-44228},
	urldate = {2024-10-18},
}
```
```yaml
nistCVE:
 author: "NIST"
 type: "Web"
 title: "CVE-2021-44228"
 url: 
    value: "https://nvd.nist.gov/vuln/detail/CVE-2021-44228"
    date: 2024-10-28
```

... it shows **"Retrieved + Date"** what I want to achieve.

<img width="903" alt="grafik" src="https://github.com/user-attachments/assets/e44035c9-e8be-457e-b491-fbe678c39953">

**However**, as you can see in the screenshot above, after the author, **the year is missing!**.
The desired output should look like this:

<img width="834" alt="grafik" src="https://github.com/user-attachments/assets/0265a115-0801-4601-9dc7-04b9e9360726">

### Steps to reproduce
- Create a project in typst, the output can be replicated in the **web application as well as locally** (I use mac + homebrew).
- Create a .bib or .yml file for the bibliography.
- Use the date (year of publication) for the reference + URL DATE (the date the site was accessed).

This can be done as follows
bibtex:
```latex
urldate = {2024-10-18},
```
yaml
```yaml
 url: 
    value: "https://nvd.nist.gov/vuln/detail/CVE-2021-44228"
    date: 2024-10-28

//OR

 url: {value: "https://nvd.nist.gov/vuln/detail/CVE-2021-44228", date: 2024-10-28}
```

Import the bibliography with APA style:

```typst
#bibliography("references.bib", title: "References", style: "apa")
```

### Versions used
Newest version of typst (Webapp)
Newest version of typst (macOS)

**Note: _I encountered this issue when I use APA style (required by the institution). When using IEEE, everyting works!_**

### Browsers

Firefox, Chrome

### OS

macOS

**Comment by mst2k** [comment]
This is actually correct APA 7 behavior. According to [APA 7 guidelines](https://apastyle.apa.org/style-grammar-guidelines/references/examples/webpage-website-references), retrieval dates should only be shown when:

- No publication date exists (shows as "n.d.")
- Content changes regularly (wikis, social media, etc.)

If you have a publication date, omit the retrieval date - this follows proper APA 7 standards.
If you need to override this behavior, modify the access macro in your CSL:

```csl
xml<macro name="access">
  <choose>
    <if variable="DOI" match="any">
      <text variable="DOI" prefix="https://doi.org/"/>
    </if>
    <else-if variable="URL">
      <group delimiter=" ">
        <choose>
          <if variable="accessed">
            <group delimiter=" ">
              <text term="retrieved" text-case="capitalize-first"/>
              <date variable="accessed" form="text" suffix=","/>
              <text term="from"/>
            </group>
          </if>
        </choose>
        <text variable="URL"/>
      </group>
    </else-if>
  </choose>
</macro>
```

This will show retrieval dates whenever the urldate field is present in your BibTeX, regardless of publication date.

**Comment by PgBiel** [comment]
This is still non-ideal output (at least the ", " there looks weird). Should we report this as an upstream bug?

**Comment by gkourgkoutas** [comment]
Thank you all for your insights on this issue!

I highlighted two points in my original post:
- When a date (e.g., 2021) is included with urldate (e.g., 2024-10-18), the output includes the year but omits “Retrieved [date]” and adds an erroneous comma -> **(2021 , )**

- When date is removed, “Retrieved [date]” appears, but the year is missing. I was expecting an output including both the year and “Retrieved [date]” without the comma, as seen in Overleaf’s LaTeX output with APA 7 (shown in my third screenshot).

@mst2k , thanks for clarifying that APA 7 guidelines omit urldate when a publication date exists, except for dynamic content. I wasn’t aware of this, so the second issue’s behavior seems correct per APA 7 standards. My expectation for including both year and urldate came from observing LaTeX output in Overleaf. Your CSL macro suggestion to force urldate display could help achieve that output, but it doesn’t address the comma bug in the first issue.

@PgBiel , you’re spot-on that the comma **(2021 , )** looks incorrect and likely a bug? Reporting this upstream seems reasonable, depending on the root cause?

**Comment by Silverarmor** [comment]
I'm confused why this hasn't gotten more attention. I'm new to typst and this breaks all `misc` bibliography entries. Bumping.

@ArttuLai 's workaround I can't get to work as there is no macro named `date-bib` in the latest CSL https://github.com/citation-style-language/styles/blob/master/apa.csl

**Comment by ArttuLai** [comment]
> there is no macro named `date-bib` in the latest CSL https://github.com/citation-style-language/styles/blob/master/apa.csl

Yes, apparently the style has changed in this regard since my post, in [PR #7510](https://github.com/citation-style-language/styles/pull/7510/changes) specifically. The macro `date-bib` is now called `date`, but the logic has changed too. The date formatting is being handled in the macro `date-issued-month-day`, while the group delimiter `, ` that caused the issue in Typst is under the `date` macro. First remove the delimiter:

```xml
<macro name="date">
  <group delimiter="-" prefix="(" suffix=")">
    <choose>
      <if variable="issued">
        <group> <!-- delimiter=", " removed -->
```

Add it back to the date:

```xml
<macro name="date-issued-month-day">
  <date variable="issued">
    <date-part name="month" prefix=", "> <!-- prefix=", " added -->
    <date-part name="day" prefix=" ">
```

As I understand it, the bug happens in Typst, where it incorrectly adds a group delimiter even when the second group is empty, e.g. when there's a year only and no specific date. This doesn't happen in entries where a year is always shown, no matter how specific date is given in the source, like `@book`. It occurs in entries like `@misc` where a year or a year-month-day can be used. I've tested with the same CSL file in Word and LaTeX, and neither exhibit the same issue.

**Comment by MathewBlanc** [comment]
I'd also like to bump this, I'm having the same issue in Quarto. I'm stuck between keeping the typst citations and manually editing the .typ file before going to pdf, or switching to citeproc and losing the citation links to the bibliography.

@ArttuLai thanks for the fix in the meantime.

**Comment by Braintelligence** [comment]
Still reproduces on hayagriva 0.10.1 (Typst 0.15.1). The access date is not involved.

`b.yaml`

```yaml
web_year:
  type: web
  title: Web page, year only
  author: Ann Author
  date: 2021
  url: https://example.org/a

web_full:
  type: web
  title: Web page, full date
  author: Ben Bookman
  date: 2021-07-14
  url: https://example.org/b

article_year:
  type: article
  title: Journal article, year only
  author: Cora Curator
  date: 2021
  parent:
    type: periodical
    title: Journal of Examples
```

`b.typ`

```typst
#bibliography("b.yaml", style: "apa", full: true)
```

Output:

```
Ann Author. (2021,). Web page, year only. https://example.org/a
Ben Bookman. (2021, July 14). Web page, full date. https://example.org/b
Cora Curator. (2021). Journal article, year only. Journal of Examples.
```

<img width="1182" height="178" alt="Image" src="https://github.com/user-attachments/assets/cca77910-4129-47c3-b1e0-b1c613a8d1c0" />

Entry 1 has no access date, so the stray comma is not caused by a publication date and an access date meeting. Entry 2 shows where the comma belongs. Entry 3 is clean because apa requests only a year for journal articles; month and day are requested for webpage entries, and their separator is emitted although they render nothing.

The same class of problem appears when a `<date>` that renders nothing sits inside a group with a delimiter:

```xml
<group delimiter="+">
  <text value="A"/>
  <date variable="issued"><date-part name="day"/></date>
  <text value="B"/>
</group>
```

For an entry whose date is `2021` this gives `A++B`, expected `A+B`. Group suppression itself works correctly: if the only variable called is empty, the whole group disappears as it should. The problem is limited to elements that are reached but produce no text.

`group_SuppressTermWhenNoOutputFromPartialDate` in #327 looks like the conformance test for this.

**Comment by jassielof** [comment]
This should be fixed with #503:

<img width="826" height="301" alt="Image" src="https://github.com/user-attachments/assets/6043b54d-fc81-4780-b9c2-176342323398" />

**Comment by Braintelligence** [comment]
I confirm your PR fixes the issue, thank you! 👍
